dali/operators/reader/loader/filesystem.cc

```diff
@@ -47,168 +47,116 @@ std::string join_path(const std::string &dir, const std::string &path) {
   // return unordered dir list. Otherwise file reader for training and validation
   // could return directories with the same names in completely different order
-  std::sort(entry_name_list.begin(), entry_name_list.end());
-  for (unsigned dir_count = 0; dir_count < entry_name_list.size(); ++dir_count) {
-      assemble_file_list(file_label_pairs, file_root, entry_name_list[dir_count], dir_count,
-                         filters, case_sensitive_filter);
-  }
-  // sort file names as well
-  std::sort(file_label_pairs.begin(), file_label_pairs.end());
-  LOG_LINE  << "read " << file_label_pairs.size() << " files from " << entry_name_list.size()
-            << "directories\n";
-
-  closedir(dir);
-
-  return file_label_pairs;
+  std::sort(subdirs.begin(), subdirs.end());
+  return true;
 }
 
 
-inline void assemble_file_list(std::vector<std::string>& file_list,
-                               const std::string &path, const std::string &curr_entry,
-                               const std::string &filter) {
-  std::string curr_dir_path = path + dir_sep + curr_entry;
-  DIR *dir = opendir(curr_dir_path.c_str());
+bool list_files(std::vector<std::string>& files, const std::string& parent_dir, 
+                std::vector<std::string> filters = {}, bool case_sensitive_filter = true) {
+
+  DIR *dir = opendir(parent_dir.c_str());
+  if (dir == nullptr)
+    return false;
 
   dirent *entry;
 
-  if (filter.empty()) {
-    while ((entry = readdir(dir))) {
+  while ((entry = readdir(dir))) {
 #ifdef _DIRENT_HAVE_D_TYPE
     /*
      * we support only regular files and symlinks, if FS returns DT_UNKNOWN
      * it doesn't mean anything and let us validate filename itself
      */
-      if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
-         entry->d_type != DT_UNKNOWN) {
-        continue;
-      }
-#endif
-      std::string rel_path = curr_entry + dir_sep + std::string{entry->d_name};
-      if (HasKnownExtension(std::string(entry->d_name))) {
-         file_list.push_back(rel_path);
-      }
+    if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
+        entry->d_type != DT_UNKNOWN) {
+      continue;
     }
-  } else {
-    // use glob to do the file search
-    glob_t pglob;
-    std::string pattern = curr_dir_path + dir_sep + filter;
-    if (glob(pattern.c_str(), GLOB_TILDE, NULL, &pglob) == 0) {
-      // iterate through the matched files
-      for (unsigned int count = 0; count < pglob.gl_pathc; ++count) {
-        std::string match(pglob.gl_pathv[count]);
-        std::string rel_path = curr_entry + dir_sep + match.substr(match.find_last_of(dir_sep)+1);
-        file_list.push_back(rel_path);
+#endif
+    std::string fname(entry->d_name);
+    for (auto &filter : filters) {
+      if (fnmatch(filter.c_str(), fname.c_str(), case_sensitive_filter ? 0 : FNM_CASEFOLD) == 0) {
+        files.emplace_back(fname);
+        break;
       }
-      // clean up
-      globfree(&pglob);
     }
   }
   closedir(dir);
```

Comment:
While we're not intentionally throwing in this function, it'd be safer to put closedir in AtScopeExit in case we add some exceptions in the future. It can theoretically leak even now in case of OOM, but I guess we don't care too much about that.